core/src/main/java/org/jdbi/v3/core/cache/internal/DefaultJdbiCache.java

```diff
@@ -0,0 +1,118 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.cache.internal;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import org.jdbi.v3.core.cache.JdbiCache;
+import org.jdbi.v3.core.cache.JdbiCacheLoader;
+
+final class DefaultJdbiCache<K, V> implements JdbiCache<K, V> {
+
+    private final ConcurrentMap<K, DoubleLinkedList.Node<K, V>> cache;
+
+    @GuardedBy("expungeQueue")
+    private final DoubleLinkedList<K, V> expungeQueue;
+
+    private final JdbiCacheLoader<K, DoubleLinkedList.Node<K, V>> cacheLoader;
+
+    private final int maxSize;
+
+    DefaultJdbiCache(DefaultJdbiCacheBuilder builder, JdbiCacheLoader<K, V> cacheLoader) {
+        this.cache = new ConcurrentHashMap<>();
+        this.expungeQueue = new DoubleLinkedList<>();
+        this.cacheLoader = wrapLoader(cacheLoader);
+
+        this.maxSize = builder.getMaxSize();
+    }
+
+    @Override
+    public V get(K key) {
+        try {
+            DoubleLinkedList.Node<K, V> node = cache.computeIfAbsent(key, cacheLoader::create);
+            refresh(node);
+            return node.value;
+        } finally {
+            expunge();
+        }
+    }
+
+    @Override
+    public V getWithLoader(K key, JdbiCacheLoader<K, V> loader) {
+        try {
+            DoubleLinkedList.Node<K, V> node = cache.computeIfAbsent(key, wrapLoader(loader)::create);
+            refresh(node);
+            return node.value;
+        } finally {
+            expunge();
+        }
+    }
+
+    @Override
+    public DefaultJdbiCacheStats getStats() {
+        synchronized (expungeQueue) {
+            return new DefaultJdbiCacheStats(expungeQueue.size, maxSize);
+        }
+    }
+
+    private JdbiCacheLoader<K, DoubleLinkedList.Node<K, V>> wrapLoader(JdbiCacheLoader<K, V> delegate) {
+
+        if (delegate == null) {
+            return k -> DoubleLinkedList.createNode(k, null);
+        }
+
+        return key -> {
+            V value = delegate.create(key);
+            DoubleLinkedList.Node<K, V> node = DoubleLinkedList.createNode(key, value);
+            if (maxSize > 0) {
+                synchronized (expungeQueue) {
+                    expungeQueue.addHead(node);
+                }
+            }
+            return node;
+        };
+    }
+
+    private void refresh(DoubleLinkedList.Node<K, V> node) {
+        if (maxSize > 0) {
+            synchronized (expungeQueue) {
+                DoubleLinkedList.Node<K, V> cacheNode = expungeQueue.removeNode(node);
+                expungeQueue.addHead(cacheNode);
+            }
+        }
```

Comment:
If an eviction raced with the reordering then the node is already unlinked by the time that the lock was acquired. This would return in an NPE when the list tries to unlink it (and would otherwise double decrement the size).

Reply:
I am not sure I understand that comment. the whole expunge procedure is protected by the same lock as the reordering. How would it race the reordering? If a thread enters expunge(), any thread trying to reorder would be held at the lock until expunging is done. If a thread enters reordering first, then the node will be atomically moved from the tail to the head and any thread trying to expunge will be held until that operation is done.

Reply:
The node is retrieved from the map outside of the `expungeQueue` lock. This means there is a race in which a read obtains the node, it is unlinked, and then the reordering is attempted.

```java
public V get(K key) {
    try {
        // threadA, time1 - obtain node
        DoubleLinkedList.Node<K, V> node = cache.computeIfAbsent(key, cacheLoader::create);
        refresh(node);
        return node.value;
    } finally {
        expunge();
    }
}

private void expunge() {
    synchronized (expungeQueue) {
        // threadB, time2 - remove node
        DoubleLinkedList.Node<K, V> node = expungeQueue.removeTail();
        if (node != null) {
            cache.remove(node.key);
}

private void refresh(DoubleLinkedList.Node<K, V> node) {
    synchronized (expungeQueue) {
        // threadA, time3 - reorder dead node
        DoubleLinkedList.Node<K, V> cacheNode = expungeQueue.removeNode(node);
        expungeQueue.addHead(cacheNode);
    }
```

Does that help explain the race?

Reply:
I see. Thanks for explaining.

Reply:
I added a fix. refresh now does nothing if the node is not linked (left or right is null) and we don't re-add it if the node was not part of the list in the first place. 

Worst case this will cause the node to be re-computed; I can see this happen only at very heavy use of the cache and the working set being just too big for the cache (so there is constant eviction pressure).